README.md
Webduino 開發板 & 公板燒錄與更新
===

此程式提供 開發板的韌體燒錄與更新，目前支援開發板如下
- ESP8266 (Smart、WeMos)
- ESP32 (Web:Bit)
- K210  (Web:AI)



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- 智慧插座燒錄工具 [#7003](https://redmine.kingkit.codes/issues/7003)
- WebEye Pro 燒錄工具 [#7004](https://redmine.kingkit.codes/issues/7004)
- smart 燒錄工具 [#6968](https://redmine.kingkit.codes/issues/6968)
- 支援 esp32-cam-mb 底板 [#6256](https://redmine.kingkit.codes/issues/6256)

[unreleased]: https://github.com/webduinoio/web-adapter/compare/v0.0.0...HEAD


esp32cam.py
import os
import time

import machine
import network
import ubinascii
import usocket

board_devSSID = "teacher001"
board_device_id = "teacher001"


def do_connect():
    global connected
    sta_if = network.WLAN(network.STA_IF)
    sta_if.active(True)
    print("connecting to network...")
    # sta_if.disconnect()
    if not sta_if.isconnected():
        sta_if.connect("WIFI_SSID", "WIFI_PASSWORD")
    cnt = 0
    while not sta_if.isconnected():
        cnt = cnt + 1
        time.sleep(0.5)
        if cnt == 60:
            break
    connected = sta_if.isconnected()
    print("network config:", sta_if.ifconfig())


class Response:
    def __init__(self, f):
        self.raw = f
        self.encoding = "utf-8"
        self._cached = None

    def close(self):
        if self.raw:
            self.raw.close()
            self.raw = None
        self._cached = None

    @property
    def content(self):
        if self._cached is None:
            try:
                self._cached = self.raw.read()
            finally:
                self.raw.close()
                self.raw = None
        return self._cached

    @property
    def text(self):
        return str(self.content, self.encoding)

    def json(self):
        import ujson

        return ujson.loads(self.content)


def request(method, url, data=None, json=None, headers={}, stream=None):
    try:
        proto, dummy, host, path = url.split("/", 3)
    except ValueError:
        proto, dummy, host = url.split("/", 2)
        path = ""
    if proto == "http:":
        port = 80
    elif proto == "https:":
        import ussl

        port = 443
    else:
        raise ValueError("Unsupported protocol: " + proto)

    if ":" in host:
        host, port = host.split(":", 1)
        port = int(port)
    # print("host:",host,",port:",port)
    ai = usocket.getaddrinfo(host, port, 0, usocket.SOCK_STREAM)
    ai = ai[0]

    s = usocket.socket(ai[0], ai[1], ai[2])
    try:
        s.connect(ai[-1])
        if proto == "https:":
            s = ussl.wrap_socket(s, server_hostname=host)
        s.write(b"%s /%s HTTP/1.0\r\n" % (method, path))
        if not "Host" in headers:
            s.write(b"Host: %s\r\n" % host)
        # Iterate over keys to avoid tuple alloc
        for k in headers:
            s.write(k)
            s.write(b": ")
            s.write(headers[k])
            s.write(b"\r\n")
        if json is not None:
            assert data is None
            import ujson

            data = ujson.dumps(json)
            s.write(b"Content-Type: application/json\r\n")
        if data:
            s.write(b"Content-Length: %d\r\n" % len(data))
        s.write(b"\r\n")
        if data:
            s.write(data)

        l = s.readline()
        # print(l)
        l = l.split(None, 2)
        status = int(l[1])
        reason = ""
        if len(l) > 2:
            reason = l[2].rstrip()
        while True:
            l = s.readline()
            if not l or l == b"\r\n":
                break
            # print(l)
            if l.startswith(b"Transfer-Encoding:"):
                if b"chunked" in l:
                    raise ValueError("Unsupported " + l)
            elif l.startswith(b"Location:") and not 200 <= status <= 299:
                raise NotImplementedError("Redirects not yet supported")
    except OSError:
        s.close()
        raise

    resp = Response(s)
    resp.status_code = status
    resp.reason = reason
    return resp


def head(url, **kw):
    return request("HEAD", url, **kw)


def get(url, **kw):
    return request("GET", url, **kw)


def post(url, **kw):
    return request("POST", url, **kw)


def put(url, **kw):
    return request("PUT", url, **kw)


def patch(url, **kw):
    return request("PATCH", url, **kw)


def delete(url, **kw):
    return request("DELETE", url, **kw)


class Res:
    def save(url, file):
        try:
            response = get(url)
            print(">>", len(response.text))
            print("get file:", file, "size:", len(response.text), ",save to:", file)
            f = open(file, "w")
            f.write(response.text)
            f.close()
            print("OK.")
        except Exception as e:
            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
            print(e)
            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")

    def get(url, file):
        try:
            response = get("https://webduinoio.github.io/pythonCode/" + url)
            print(">>", len(response.text))
            print("get file:", file, "size:", len(response.text), ",save to:", file)
            f = open(file, "w")
            f.write(response.text)
            f.close()
            print("OK.")
        except Exception as e:
            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
            print(e)
            print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")

    def exe(dir):
        srcDir = dir
        try:
            while True:
                idx = dir.index("/")
                try:
                    name = dir[0:idx]
                    try:
                        print("mkdir", name)
                        os.mkdir(name)
                    except:
                        pass
                    try:
                        os.chdir(name)
                        print("cd", name)
                    except:
                        pass
                except:
                    pass
                dir = dir[idx + 1 :]
        except:
            pos = -1
            try:
                pos = dir.index(".mpy")
            except:
                pass
            try:
                if pos == -1:
                    pos = dir.index(".py")
            except:
                pass
            try:
                if pos > 0:
                    pyFile = dir
                    Res.get(srcDir, pyFile)
                else:
                    try:
                        print("mkdir", dir)
                        os.mkdir(dir)
                    except:
                        pass
                    try:
                        os.chdir(dir)
                        print("cd ", dir)
                    except:
                        pass
            except:
                pass
        os.chdir("/")


def install(deviceId=""):

    # WiFi Connect
    print("connect...")
    do_connect()
    print("get files...")

    # 開源必備
    Res.exe("lib/urequests.py")
    Res.exe("lib/umqtt/simple.py")

    # Webduino 類別庫
    Res.exe("lib/webduino/led.py")
    Res.exe("lib/webduino/config.py")
    Res.exe("lib/webduino/gdriver.py")
    Res.exe("lib/webduino/camera.py")
    Res.exe("lib/webduino/board.py")
    Res.exe("lib/webduino/mqtt.py")
    Res.exe("lib/webduino/wifi.py")
    Res.exe("lib/webduino/webserver.py")
    Res.exe("lib/webduino/debug.py")
    Res.exe("lib/utils.py")  # save url to file
    Res.get("", "index.html")

    from utils import Utils
    from webduino.config import Config

    Utils.save("https://webduinoio.github.io/pythonCode/app/boot.py", "boot.py")
    Utils.save("https://webduinoio.github.io/pythonCode/app/CamApp.py", "main.py")
    Config.load()
    if not deviceId == "":
        Config.data["devId"] = deviceId
    else:
        deviceId = Config.data["devId"]

    if not board_devSSID == "":
        Config.data["devSSID"] = board_devSSID

    print("-=-=-=-=-=-=-=-=-=-=-=-=-=-=-")
    print("-    Device ID: [" + deviceId + "]    -")
    Config.save()
    print("-=-=-=-=-=-=-=-=-=-=-=-=-=-=-")
    print("Mac address:", ubinascii.hexlify(network.WLAN().config("mac"), ":").decode())


install(deviceId=board_device_id)
time.sleep(1)
machine.reset()


lib/repl.js
const generateESP32UploadCode = (file, pythonCode) => {
  code = `
pycode = """
${pythonCode}
"""
import machine
# write python code
with open('${file}', 'w') as f:
    f.write(pycode)
with open ('${file}', 'r') as f:
    content = f.read()
`;
  return code;
};

const generateUploadCode = (type, file, pythonCode) => {
  code = `
pycode = """
${pythonCode}
"""
from Maix import utils
from time import sleep
import gc, machine, ujson
# write python code
with open('${file}', 'w') as f:
    f.write(pycode)
sleep(0.5)
with open ('${file}', 'r') as f:
    content = f.read()
# save firmware type
romFlagAddressStart = 0x1FFFF
preFlag = int.from_bytes(utils.flash_read(romFlagAddressStart, 1), "big")
romFlag = 1 if "${type}" == "mini" else 0
if preFlag != romFlag:
    utils.flash_write(romFlagAddressStart, bytes([romFlag]))
deployCmd = '_DEPLOY/{"url":"local"}'
cfg.init()
cfg.put('cmd', deployCmd)
`;
  return code;
};

const snapshotESP32Code = `
import camera,sys
if not 'cameraStatus' in locals():
  camera.init(0, format=camera.JPEG,xclk_freq=camera.XCLK_20MHz)
  camera.framesize(15)
  camera.framesize(camera.FRAME_SVGA)
  camera.quality(10)
  cameraStatus = True
jpg = camera.capture()
sys.stdout.write("JPGSize:")
sys.stdout.write(str(len(jpg)))
sys.stdout.write('\\r\\n')
`;

const snapshotK210Code = `
from webai import *
from time import sleep
webai.init()
webai.lcd.init()
repl = UART.repl_uart()
img = webai.snapshot()
webai.show(img=img)
jpg = img.compress(80)
img = None
jpg = jpg.to_bytes()
repl.write("JPGSize:")
repl.write(str(len(jpg)))
repl.write('\\r\\n')
`;

class DataTransformer {
  constructor() {
    this.container = "";
    this.decoder = new TextDecoder();
    this.readLine = true;
    this.readByteArray = false;
  }

  setReadLine() {
    this.readLine = true;
    this.readByteArray = false;
  }

  setReadByteArray(bytes) {
    this.readLine = false;
    this.readByteArray = true;
    this.byteArray = new Uint8Array();
    this.OK = false;
  }

  transform(chunk, controller) {
    if (this.readLine) {
      chunk = this.decoder.decode(chunk);
      this.container += chunk;
      const lines = this.container.split("\r\n");
      this.container = lines.pop();
      lines.forEach((line) => controller.enqueue(line));
    }
    if (this.readByteArray) {
      this.byteArray = new Uint8Array([...this.byteArray, ...chunk]);
      var byteArrayLength = this.byteArray.length;
      // check endof ArrayBuffer 0x04 , 0x04 , 62
      var endSign = chunk.slice(chunk.length - 3);
      if (endSign[0] == 4 && endSign[1] == 4 && endSign[2] == 62) {
        var allData = new Uint8Array([
          ...this.byteArray.slice(0, byteArrayLength - 3),
        ]);
        var rtnData = [];
        var start = false;
        // find startPos '>OK'
        for (var i = 0; i < allData.length; i++) {
          if (!start) {
            if (
              allData[i] == 0x3e &&
              allData[i + 1] == 0x4f &&
              allData[i + 2] == 0x4b
            ) {
              i = i + 2;
              start = true;
              continue;
            }
          } else {
            if (allData[i] == 0x0d && allData[i + 1] == 0x0a) {
              continue;
            } else {
              rtnData.push(allData[i]);
            }
          }
        }
        rtnData = new Uint8Array(rtnData);
        controller.enqueue(rtnData);
      }
    }
  }

  flush(controller) {
    controller.enqueue(this.container);
  }
}

class REPL {
  constructor() {
    this.encoder = new TextEncoder();
    this.decoder = new TextDecoder();
    this.callback = function () {};
  }

  addListener(callback) {
    this.callback = callback;
  }

  async usbConnect(board = "esp32cam") {
    var self = this;
    const filter = { usbVendorId: 6790 };
    if (self.port == undefined) {
      self.port = await navigator.serial.requestPort({
        /*filters: [filter]*/
      });
      await this.port.open({ baudRate: 115200, dateBits: 8, stopBits: 1 });
      if (board == "esp32cam") {
        await this.port.setSignals({ dataTerminalReady: false });
        await this.port.setSignals({ requestToSend: false });
      }
      this.writer = this.port.writable.getWriter();
      this.stream = new DataTransformer();
      this.reader = this.port.readable
        .pipeThrough(new TransformStream(this.stream))
        .getReader();
      self.port.ondisconnect = function () {
        console.log("disconnect port");
        self.port = null;
      };
    }
  }

  async restart(board) {
    if (board == "esp32cam") {
      await this.waitRestart();
    } else {
      for (var i = 0; i < 30; i++) {
        await this.port.setSignals({ dataTerminalReady: false });
        await new Promise((resolve) => setTimeout(resolve, 10));
        await this.port.setSignals({ dataTerminalReady: true });
      }
      if (board != "esp32") {
        await new Promise((resolve) => setTimeout(resolve, 1500));
      } else {
        console.log("esp32 restart");
      }
      console.log("restart ok");
    }
  }

  async waitRestart() {
    let value, done;
    while (1) {
      await new Promise((resolve) => setTimeout(resolve, 10));
      await this.writer.write(Int8Array.from([0x03 /*interrupt*/]));
      ({ value, done } = await this.reader.read());
      // console.log("value:" + value+",done:"+done+",len:"+value.length);
      if (value.substring(0, 35) === 'Type "help()" for more information.') {
        break;
      }
    }
    console.log("restart ok");
  }

  async enter(board) {
    await this.restart(board);
    for (var i = 0; i < 100; i++) {
      await new Promise((resolve) => setTimeout(resolve, 10));
      await this.writer.write(Int8Array.from([0x03 /*interrupt*/]));
    }
    let { value, done } = await this.reader.read();

    console.log("clear boot data");
  }

  async write(code, cb) {
    var boundry = "===" + Math.random() + "==";
    await this.writer.write(Int8Array.from([0x01 /*RAW paste mode*/]));
    code = "\r\nprint('" + boundry + "')\r\n" + code;
    code = code + "\r\nprint('" + boundry + "')\r\n";
    await this.writer.write(this.encoder.encode(code));
    await this.writer.write(Int8Array.from([0x04 /*exit*/]));
    var startBoundry = false;
    var rtnObj = "" + code.length;
    while (true) {
      var { value, done } = await this.reader.read();
      if (this.stream.readLine) {
        if (value == ">OK" + boundry) {
          //console.log("startBoundry....");
          startBoundry = true;
          continue;
        }
        if (value == boundry) {
          //console.log("endBoundry , rtnObj:", rtnObj);
          return rtnObj;
        } else if (startBoundry && cb != null) {
          var { value, done } = await cb(value);
          if (done) return value;
        }
      }
      if (this.stream.readByteArray) {
        var { value, done } = await cb(value, true);
        if (done) return value;
      }
    }
  }

  async uploadFile(type, filename, pythonCode) {
    if (type == "esp32") {
      pythonCode = generateESP32UploadCode(filename, pythonCode);
      pythonCode = pythonCode.replaceAll("\\", "\\\\");
      var rtn = await this.write(pythonCode, function (value) {
        if (value.substring(0, 4) == "save") {
          return { value: value, done: true };
        }
      });
      return rtn;
    } else {
      pythonCode = generateUploadCode(type /*std|mini*/, filename, pythonCode);
      pythonCode = pythonCode.replaceAll("\\", "\\\\");
      return await this.write(pythonCode, function (value) {
        if (value.substring(0, 4) == "save") {
          return { value: value, done: true };
        }
      });
    }
  }

  async setWiFi(pythonCode, ssid, pwd) {
    pythonCode += "cfg.init()\n";
    pythonCode +=
      "cfg.put('wifi',{'ssid':'" + ssid + "','pwd':'" + pwd + "'})\n";
    return await this.write(pythonCode, function (value) {
      if (value.substring(0, 4) == "save") {
        return { value: value, done: true };
      }
    });
  }

  async snapshot(board) {
    this.stream.setReadLine();
    var code = board == "k210" ? snapshotK210Code : snapshotESP32Code;
    //console.log(">>>", code);
    var imgSize = await this.write(code, async function (value) {
      console.log("value>>", value);
      if (value.substring(0, 8) == "JPGSize:") {
        value = parseInt(value.substring(8));
        return { value: value, done: true };
      } else {
        return { value: value, done: false };
      }
    });
    await this.reader.read();
    console.log("size:", imgSize);
    await this.writer.write(Int8Array.from([0x01 /*RAW paste mode*/]));
    var cmd = board == "k210" ? "repl.write(jpg)" : "sys.stdout.write(jpg)";
    await this.writer.write(this.encoder.encode(cmd));
    await this.writer.write(Int8Array.from([0x04 /*exit*/]));
    var urlCreator = window.URL || window.webkitURL;
    this.stream.setReadByteArray(parseInt(imgSize));
    var { value, done } = await this.reader.read();
    var jpg = new Blob([value], { type: "image/jpeg" });
    return urlCreator.createObjectURL(jpg);
  }
}


lib/esptool/esptool.js
// Base on https://github.com/espressif/esptool/tree/746023b5dbed058ddc8cdf79f5de5b15e76c401e
// Protocol: https://github.com/espressif/esptool/wiki/Serial-Protocol#initial-synchronisation 

let ESP = (function () {
  // Commands supported by ESP8266 ROM bootloader
  const ESP_REQUEST = 0x0
  const ESP_RESPOND = 0x1
  const ESP_FLASH_BEGIN = 0x02
  const ESP_MEM_BEGIN = 0x05
  const ESP_MEM_END = 0x06
  const ESP_MEM_DATA = 0x07
  const ESP_SYNC = 0x08
  const ESP_WRITE_REG = 0x09
  const ESP_READ_REG = 0x0a

  // OTP ROM addresses
  const ESP_OTP_MAC0 = 0x3ff00050
  const ESP_OTP_MAC1 = 0x3ff00054
  const ESP_OTP_MAC2 = 0x3ff00058
  const ESP_OTP_MAC3 = 0x3ff0005c

  const SPI_REG_BASE = 0x60000200
  const SPI_USR_OFFS = 0x1c
  const SPI_USR1_OFFS = 0x20
  const SPI_USR2_OFFS = 0x24
  const SPI_W0_OFFS = 0x40

  // Maximum block sized for RAM and Flash writes, respectively.
  const ESP_RAM_BLOCK = 0x1800
  const FLASH_WRITE_SIZE = 0x4000


  const ESP_SPI_SET_PARAMS = 0x0B
  const ESP_SPI_ATTACH = 0x0D
  const ESP_READ_FLASH_SLOW = 0x0e  // ROM only, much slower than the stub flash read
  const ESP_CHANGE_BAUDRATE = 0x0F
  const ESP_FLASH_DEFL_BEGIN = 0x10
  const ESP_FLASH_DEFL_DATA = 0x11
  const ESP_FLASH_DEFL_END = 0x12
  const ESP_SPI_FLASH_MD5 = 0x13

  // Initial state for the checksum routine
  const ESP_CHECKSUM_MAGIC = 0xef

  const UART_CLKDIV_REG = 0x60000014
  const UART_DATE_REG_ADDR = 0x60000078  // used to differentiate ESP8266 vs ESP32*
  const UART_DATE_REG2_ADDR = 0x3f400074 // used to differentiate ESP32-S2 vs other models

  // Some commands supported by stub only
  const ESP_ERASE_FLASH = 0xD0
  const ESP_ERASE_REGION = 0xD1
  const ESP_READ_FLASH = 0xD2
  const ESP_RUN_USER_CODE = 0xD3

  // CMD
  const ESP_CMD_NOT_IMPLEMENTED = 0xFF
  //const ESP_SYNC_BAUD = 115200
  const ESP_SYNC_BAUD = 115200 * 4

  const ESPInfo = {
    ESP8266: {
      TEXT_MEM_OFFSET: 0x4010E000,
      DATA_MEM_OFFSET: 0x3FFFACA8,
      ENTRY_ADDR: 0x4010E004,
      DATE_REG_VALUE: 0x00062000,
      DATE_REG2_VALUE: null
    },
    ESP32: {
      TEXT_MEM_OFFSET: 0x400BE000,
      DATA_MEM_OFFSET: 0x3FFDEBA8,
      ENTRY_ADDR: 0x400BE594,
      DATE_REG_VALUE: 0x15122500,
      DATE_REG2_VALUE: null
    },
    ESP32S2: {
      TEXT_MEM_OFFSET: 0x40028000,
      DATA_MEM_OFFSET: 0x3ffe2bf4,
      ENTRY_ADDR: 0x4002873C,
      DATE_REG_VALUE: 1280,
      DATE_REG2_VALUE: null
    }
  }

  let syncID = null
  let et = new EventTarget()
  let decoder = new slip.Decoder({
    onMessage: parseMsg,
    maxMessageSize: 209715200,
    bufferSize: 20000
  })



  class ESP extends WebSerial {
    constructor(options, logger) {
      super();
      this.options = options;
      this.chip = null;
      ESP.console = typeof logger == 'undefined' ? console : logger;
    }
    /**
     * 按鈕觸發init，初始化接收與發送的功能，連接時baud為115200最為穩定
     */
    async init() {
      // this.port = await super.connect({ baudrate: ESP_SYNC_BAUD })
      if (this.port == null) {
        this.port = await super.connect(this.options);
        this.on("data", (chunk) => {
          decoder.decode(chunk);
        });
      }
    }

    toUartBootMode() {
      return new Promise(async resolve => {
        ESP.console.log("To UART Bootloader Mode")
        let i = 5
        // dtr: true => IO0 = LOW
        // rts: true => chip in reset
        while (i > 0) {
          await this.port.setSignals({ dataTerminalReady: false, requestToSend: true })
          await delay(0.1)
          await this.port.setSignals({ dataTerminalReady: true, requestToSend: false })
          await delay(0.5)
          await this.port.setSignals({ dataTerminalReady: false })
          i--
        }
        ESP.console.log("UART Bootloader Mode success")
        resolve()
      })
    }

    async sync() {
      return new Promise(resolve => {
        ESP.console.log("sync...")
        let syncCmd = [ESP_REQUEST, ESP_SYNC, 0x24, 0x0, 0x0, 0x0, 0x0, 0x0, 0x07, 0x07, 0x12, 0x20]
        for (let i = 0; i < 32; i++) {
          syncCmd.push(0x55)
        }
        let packet = slip.encode(new Uint8Array(syncCmd))
        let syncEvent = ESP_SYNC.toString(16)

        et.addEventListener(syncEvent, function done() {
          et.removeEventListener(syncEvent, done)
          resolve()
        })

        syncID = setInterval(() => {
          this.write(packet)
          ESP.console.log(".")
        }, 500)
      })
    }
    getChipType() {
      ESP.console.log("Detecting chip type...")
      return new Promise(async resolve => {
        let uartRegAddr = await this.req(ESP_READ_REG, { UART_DATE_REG_ADDR })
        await this.req(ESP_READ_REG, { UART_DATE_REG2_ADDR })
        console.log(">>> uartRegAddr >>>>",uartRegAddr);
        for (let field of ['ESP8266', 'ESP32', 'ESP32S2']) {
          if (uartRegAddr === ESPInfo[field].DATE_REG_VALUE) {
            this.chip = field
            console.log(">>> chip >>>>",field);
            break
          }
        }
        if (this.chip === null) {
          ESP.console.log("Chip is not supported !")
        }
        resolve()
      })
    }
    getCrystal() {
      return new Promise(async resolve => {
        ESP.console.log("Crystal is XX")
        await this.req(ESP_READ_REG, { UART_CLKDIV_REG })
        resolve()
      })
    }
    getMAC() {
      return new Promise(async resolve => {
        ESP.console.log("getMAC...")
        let mac0 = await this.req(ESP_READ_REG, { address: ESP_OTP_MAC0 })
        let mac1 = await this.req(ESP_READ_REG, { address: ESP_OTP_MAC1 })
        let mac3 = await this.req(ESP_READ_REG, { address: ESP_OTP_MAC3 })
        let oui

        if (mac3 != 0) {
          oui = [(mac3 >> 16) & 0xff, (mac3 >> 8) & 0xff, mac3 & 0xff]
        }
        else if (((mac1 >> 16) & 0xff) == 0) {
          oui = [0x18, 0xfe, 0x34]
        }
        else if (((mac1 >> 16) & 0xff) == 1) {
          oui = [0xac, 0xd0, 0x74]
        }
        let mac = oui.concat([(mac1 >> 8) & 0xff, mac1 & 0xff, (mac0 >> 24) & 0xff])
        mac = mac.map(e => e.toString(16)).toString().replace(/,/g, ":")

        ESP.console.log(`MAC: ${mac}`)
        resolve()
      })
    }
    uploadStub() {
      return new Promise(async resolve => {
        if (this.chip === null) resolve()
        ESP.console.log(this.chip)
        let stub = {}
        stub['text_start'] = ESPInfo[this.chip].TEXT_MEM_OFFSET
        stub['data_start'] = ESPInfo[this.chip].DATA_MEM_OFFSET

        ESP.console.log("Uploading stub...")

        for (let field of ['text', 'data']) {
          await fetch(`./lib/esptool/stub/${this.chip}/${field}.stub`)
            .then(res => {
              return res.arrayBuffer()
            }).then(buf => {
              stub[field] = buf
            })

          let offs = stub[field + '_start']
          let length = stub[field].byteLength
          let blocks = Math.ceil(length / ESP_RAM_BLOCK)

          // ESP.console.log("mem begin")

          await this.req(ESP_MEM_BEGIN,
            {
              size: length,
              blocks: blocks,
              blockSize: ESP_RAM_BLOCK,
              offset: offs
            })

          for (let seq = 0; seq < blocks; seq++) {
            // ESP.console.log("seq", seq)
            let from_offs = seq * ESP_RAM_BLOCK
            let to_offs = from_offs + ESP_RAM_BLOCK
            let data = stub[field].slice(from_offs, to_offs)
            // ESP.console.log(data.length)
            // ESP.console.log("mem data")
            // ESP.console.log("dataSize:", data.byteLength)

            let chk = checksum(data)
            // ESP.console.log("chk:", chk)
            await this.req(ESP_MEM_DATA,
              {
                dataSize: data.byteLength,
                sequence: seq,
                zero1: 0,
                zero2: 0,
                file: data
              },
              chk/* checksum */)
          }
        }

        ESP.console.log("Running stub...")

        // ESP.console.log("mem finish")
        await this.req(ESP_MEM_END, { executeFlag: 0, entryAddress: ESPInfo[this.chip].ENTRY_ADDR })

        resolve()
      })
    }
    changeBaudRate(baudrate) {
      return new Promise(async resolve => {
        ESP.console.log("Board is changing baud to", baudrate)
        await this.req(ESP_CHANGE_BAUDRATE, {
          burnBaud: baudrate,
          syncBaud: ESP_SYNC_BAUD
        })
        ESP.console.log("Changed.")
        await delay(0.05)

        await this.setBaudRate(baudrate)
        resolve()
      })
    }
    eraseFlash() {
      return new Promise(async resolve => {
        ESP.console.log("Erasing flash (this may take a while)...")
        await this.req(ESP_ERASE_FLASH)

        ESP.console.log("Chip erase completed successfully !")
        resolve()
      })
    }
    async writeFlash(address, url, progressFn) {
      return new Promise(async (resolve, reject) => {
        ESP.console.log("start write flash")
        let image
        await fetch(url, {
          mode: 'cors'
        }).then(res => {
          ESP.console.log(res)
          return res.arrayBuffer()
        }).then(buf => {
          image = buf
        })
        let uncsize = image.byteLength
        let calcmd5 = md5(image)
        // ESP.console.log("file md5:", calcmd5)

        // compress image
        image = pako.deflate(image, { level: 9 })
        let size = image.byteLength
        let ratio = uncsize / size

        // compsize
        let num_blocks = Math.floor((size + FLASH_WRITE_SIZE) / FLASH_WRITE_SIZE)

        let write_size = uncsize
        ESP.console.log(`Compressed ${uncsize} bytes to ${size}...`)

        // enter compressed flash mode
        // ESP.console.log("flash begin")
        await this.req(ESP_FLASH_DEFL_BEGIN, {
          write_size,
          num_blocks,
          FLASH_WRITE_SIZE,
          address,
        })

        // argfile.seek(0)  # in case we need it again
        let seq = 0
        let written = 0
        while (image.byteLength > 0) {

          let addressIndex = toHEX(address + seq * FLASH_WRITE_SIZE, 8)
          ESP.console.log(`Writing at 0x${addressIndex}... (${Math.floor(100 * (seq + 1) / num_blocks)}%)`)
          progressFn(`${Math.floor((100 * (seq + 1)) / num_blocks)}`);
          let block = image.slice(0, FLASH_WRITE_SIZE)

          let chk = checksum(block)
          await this.req(ESP_FLASH_DEFL_DATA,
            {
              size: block.byteLength,
              sequence: seq,
              zero1: 0,
              zero2: 0,
              file: block
            },
            chk)
          image = image.slice(FLASH_WRITE_SIZE)
          seq += 1
          written += block.byteLength
        }
        ESP.console.log("flash end")
        await this.verify(address, uncsize, calcmd5)
        .then((result) => {
          resolve(result);
        })
          .catch((error) => {
          reject(error);
        });
      })
    }
    async verify(address, size, fileHash) {
      return new Promise(async (resolve, reject) => {
        let hash = await this.req(ESP_SPI_FLASH_MD5, {
          address,
          size,
          padding: 0,
          padding2: 0
        })
        if (hash === fileHash) {
          ESP.console.log("verify ok")
          resolve(1);
        } else {
          ESP.console.log("md5 hash error !")
          reject(0);
        }
      });
    }
    async setSPI() {
      return new Promise(async resolve => {
        ESP.console.log("set spi...")
        await this.req(ESP_SPI_SET_PARAMS, {
          id: 0,
          totalSize: 0x00400000,
          blockSize: 0x00010000,
          sectorSize: 0x00001000,
          pageSize: 0x00000100,
          statusMask: 0x0000ffff,
        })
        resolve()
      })
    }
    /**
     * @param {string, array} fw firmware 
     */
    async burn(fw, progressFn = () => { }) {
      return new Promise(async (resolve,reject) => {
        await this.toUartBootMode();
        await this.sync();
        await this.getChipType();
        await this.uploadStub();

        // await this.changeBaudRate(this.options.baudrate)

        if (Array.isArray(fw)) {
          for (let [fwUrl, addr] of fw) {
            // await this.setSPI()
            await this.writeFlash(addr, fwUrl, progressFn);
          }
        } else if (typeof fw === "string") {
          await this.writeFlash(0x0, fw, progressFn)
            .then((result) => {
              resolve(result);
            })
            .catch((error) => {
              reject(error);
            });
        } else {
          ESP.console.log("Firmware format is wrong");
        }

        // 0x13 SPI_FLASH_MD5 => Hash of data verified.
        // 0x02 FLASH_BEGIN => Leaving...
        // 0x12
        await this.req(ESP_FLASH_DEFL_END, {
          reboot: 1
        })
        ESP.console.log("Reboot...")
        await this.reset()
      });
    }

    async erase() {
      return new Promise(async (resolve,reject) => {

        try {
          await this.toUartBootMode()
          await this.sync()
          await this.getChipType()
          await this.uploadStub()
          await this.eraseFlash()
          this.reset()
          resolve();
        } catch (error) {
          reject();
        }
      })
    }
    async reset() {
      return new Promise(async resolve => {
        ESP.console.log("reset...")
        await this.port.setSignals({ requestToSend: true })
        await delay(0.1)
        await this.port.setSignals({ requestToSend: false })
        resolve()
      })
    }
    stop() {
      clearTimeout(syncID)
    }
    /**
    * Requeset Command
    * @param {number} op
    * @param {number} data
    * @param {number} checksum
    */
    req(op, data, checksum = 0) {

      let dataSize = 0
      if (typeof data === "object") {
        for (let x in data) {
          if (x !== 'file') {
            dataSize += 4
          } else {
            dataSize += data.file.byteLength
          }
        }
      }

      // console.log("op", op)
      // console.log("dataSize", dataSize)
      // console.log("checksum", checksum)
      // console.log("data", data)

      // request + opcode + size + checksum = 1 + 1 + 2 + 4 = 8
      let command = new ArrayBuffer(dataSize + 8)
      let view = new DataView(command, 0)
      let pos = 0

      //request
      view.setInt8(pos, ESP_REQUEST, true /* little endian */)
      pos++

      // opcode
      view.setInt8(pos, op, true)
      pos++

      // size
      view.setInt16(pos, dataSize, true)
      pos = pos + 2

      // checksum
      view.setInt32(pos, checksum, true)
      pos = pos + 4

      for (let x in data) {
        if (x === "file") {
          let file = new Uint8Array(data['file'])
          for (let i = 0; i < file.byteLength; i++) {
            view.setInt8(pos, file[i], false) // no litte endian
            pos++
          }

        } else {
          view.setInt32(pos, data[x], true)
          pos = pos + 4
        }
      }

      // ESP.console.log("command", command)
      let packet = slip.encode(command)
      this.write(packet)

      let opEvent = op.toString(16)
      return new Promise(resolve => {
        et.addEventListener(opEvent, function done(e) {
          et.removeEventListener(opEvent, done)
          resolve(e.detail)
        })
      })
    }
  }

  function parseMsg(msg) {
    // ESP.console.log("msg", msg)

    if (msg[0] == ESP_RESPOND) {
      let op = msg[1]
      let packetSize = msg[2] | msg[3] << 8
      let value = msg[4] | (msg[5] << 8) | (msg[6] << 16) | (msg[7] << 24)
      // ESP.console.log(op, packetSize, value)

      let opEvent = new Event(op.toString(16))

      switch (op) {
        case ESP_SYNC:
          if ((msg[4] == 0x07 && msg[5] == 0x07 && msg[6] == 0x12 && msg[7] == 0x20) || msg[8] == ESP_CMD_NOT_IMPLEMENTED) {
            clearTimeout(syncID)
            et.dispatchEvent(opEvent)
          }
          break
        case ESP_READ_REG:
          opEvent = new CustomEvent(ESP_READ_REG.toString(16), {
            detail: value
          })
          et.dispatchEvent(opEvent)
          break
        case ESP_CHANGE_BAUDRATE:
          et.dispatchEvent(opEvent)
          break
        case ESP_SPI_SET_PARAMS:
          ESP.console.log("get spi:", msg)
          et.dispatchEvent(opEvent)
          break
        case ESP_SPI_FLASH_MD5:
          let hash = "";
          for (let i = 8; i < 24; i++) {
            hash += toHEX(msg[i], 2)
          }
          // ESP.console.log("hash:", hash)
          opEvent = new CustomEvent(ESP_SPI_FLASH_MD5.toString(16), {
            detail: hash
          })
          et.dispatchEvent(opEvent)
          break
        default:
          et.dispatchEvent(opEvent)
      }
    } else if (String.fromCharCode(...msg) === "OHAI") {
      ESP.console.log("Stub running...")
    }
  }

  function checksum(data, state = ESP_CHECKSUM_MAGIC) {
    for (let b of new Uint8Array(data)) {
      state ^= b
    }
    return state
  }

  /**
   * Reverse every two character 
   * @param {string} num
   * @returns {string}
   */
  function reverseString(str) {
    return str.match(/.{2}/g).reverse().join("")
  }

  /**
  * Convert to Hex String and Pad Zero
  * @param {number} num
  * @param {number} length
  * @returns {string}
  */
  function toHEX(num, length) {
    let str = num.toString(16)
    let len = str.length
    while (len < length) {
      str = "0" + str
      len = str.length
    }
    return str
  }

  function delay(time) {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve()
      }, time * 1000)
    })
  }

  return ESP
})()